### PR TITLE
Game saving

### LINKED ISSUE #30: Saving and loading
The gooey is almost complete. I can complete all mini-backend functions by today. The only remaining TODO is documentation and saving/loading.

### PR DESCRIPTION
**Pull request by vincentjguo**
Adds game saving and retrieval. Resolves #30

### CODE REVIEW

**Comment by potatoeggy** [comment]
I would do this from MainWindow instead, as GameWindow is to be kept as backend-free as possible. Additionally, if you're applying changes from the `ai_gui` branch, please make a pull request off of that instead of rebasing.

**Comment by vincentjguo** [comment]
I can't access the variables from gamewindow from another class. If I could I would've done all the processing in the save class.

**Comment by potatoeggy** [comment]
Which is why I'm asking you to put it in MainWindow instead, since it has access to variables. MainWIndow is also a minibackend at this point, anyway.

**Comment by bob2957** [comment]
Alright
Vincent Guo
Sergeant
Cyclone Flight Supply NCO
8 Globemaster Squadron
Royal Canadian Air Cadets
vincentguo04@gmail.com
6476412478

**Comment by potatoeggy** [comment]
let's get rid of that shall we

**Comment by vincentjguo** [comment]
wat

**Comment by potatoeggy** [comment]
you had a bit of personal information in your signature

**Comment by vincentjguo** [comment]
ah was it my cadet signature?

File: src/AI.java [hunk 1]
```diff
@@ -29,6 +29,7 @@ public int compareTo(coordinate o) {
 public class AI {
 
 	static Board b = new Board();
+	static ArrayList<Integer> fullCol = new ArrayList<>(); //keeps track of full columns
```
Comment (potatoeggy) [inline src/AI.java:32]: This re-enables pruning; is this intentional?

File: src/GameWindow.java [hunk 2]
```diff
@@ -207,4 +234,44 @@ public Board getBoard() {
...
+			for (int j = 0; j < board.W; j++) {
+				board.board[i][j] = input.nextInt();
+			}
+			input.nextLine();
+		}
+		currentPlayer = input.nextInt();
+		cpuDifficulty = input.nextInt();
+		players = new int[]{input.nextInt(), input.nextInt()};
+		input.nextLine();
+		names = new String[]{input.next(), input.next()};
```
Comment (potatoeggy) [inline src/GameWindow.java:272]: "Player 1" and "Player 2" (default names) are going to be cut off

File: src/GameWindow.java [hunk 3]
```diff
@@ -207,4 +234,44 @@ public Board getBoard() {
...
+		internalTurnCount = input.nextInt();
+		buttonsFilled = input.nextInt();
+		actionLock = input.nextBoolean();
```
Comment (potatoeggy) [inline src/GameWindow.java:275]: Action lock doesn't need to be saved, sorry

File: src/Save.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.PrintWriter;
+
+public class Save {
```
Comment (potatoeggy) [inline src/Save.java:5]: I'd recommend passing a MainWindow parameter to gain access to GameWindow variables.

**Review by potatoeggy — COMMENTED** [review]
In general, please move code out of GameWindow and instead dump it in MainWindow or another class. If using the latter, very little code should be needed in the other classes (only MainWindow should have a couple of lines)

**Comment by vincentjguo** [comment]
Ok, im going to make a new branch, but don't delete this one - I need it to copy and paste.

**Comment by potatoeggy** [comment]
`git checkout`

**Comment by vincentjguo** [comment]
I know but this branch is based on an old version of `master`

**Pull request closed without merging**